lib/types-simple.nix

```diff
@@ -0,0 +1,457 @@
+
+
+  ## -- TYPE SETUP STUFF --
+
+  mkBaseType = {
+    # unique name (for matching on the type)
+    name,
+    # the (displayable) type description
+    description,
+    # a function to check the outermost type, given a value (Val -> Bool)
+    # TODO: this is value-specific, maybe this should be inside the type checker
+    # logic instead of the type definiton? There’s some repetition.
+    check,
+    # the variant of this type
+    variant,
+    # extra fields belonging to the variant
+    extraFields
+  }: { inherit name description check variant; } // extraFields;
+
+  mkScalar = { name, description, check }: mkBaseType {
+    inherit name description check;
+    variant = variants.scalar;
+    extraFields = {};
+  };
+
+  mkRecursive = { name, description, check,
+    # return all children for a value of this type T t,
+    # give each child (of type t) a displayable name.
+    # (T t -> Map Name t)
+    each,
+    # The nested value t of the type functor
+    nested
+  }: mkBaseType {
+    inherit name description check;
+    variant = variants.recursive;
+    extraFields = { inherit each nested; };
+  };
+
+
+  ## -- TYPES --
+
+  # the type with no inhabitants (kind of useless …)
+  void = mkScalar {
+    name = "void";
+    description = "void";
+    # there are no values of type void
+    check = lib.const false;
+  };
+
+  # the any type, every value is an inhabitant
+  # it basically turns off the type system, use with care
+  any = mkScalar {
+    name = "any";
+    description = "any type";
+    check = lib.const true;
+  };
+
+  # the type with exactly one inhabitant
+  unit = mkScalar {
+    name = "unit";
+    description = "unit";
+    # there is exactly one unit value, we represent it with {}
+    # Q: why not `null`?
+    # A: `null` has strong connotations as the “always existing”
+    # alternative value; of course in a unityped language like
+    # nix this is moot, but here we take the chance to throw out
+    # this harmful idea (the “million dollar mistake”).
+    check = v: v == {};
+  };
+
+  # the type with two inhabitants
+  bool = mkScalar {
+    name = "bool";
+    description = "boolean";
+    check = builtins.isBool;
+  };
+
+  # a nix string
+  string = mkScalar {
+    name = "string";
+    description = "string";
+    check = builtins.isString;
+  };
+
+  # a signed nix integer
+  int = mkScalar {
+    name = "int";
+    description = "integer";
+    check = builtins.isInt;
+  };
+
+  # a nix floating point number
+  float = mkScalar {
+    name = "float";
+    description = "float";
+    check = builtins.isFloat;
+  };
+
+  # helper for descriptions of recursive types
+  # TODO: descriptions need to assume t is a type,
+  # which is only true for Fix Type. How to make nice?
+  describe = t: t.description or "<non-type>";
+
+  # list with children of type t
+  # list bool:
+  #   [ true false false ]
+  # list (attrs unit):
+  #   [ { a = {}; } { b = {}; } ]
+  #   []
+  list = t: mkRecursive {
```

Comment:
Maybe call this `listOf` so we use the same naming as with NixOS modules.

Reply:
I deliberately changed the names to be internally consistent; the module system has `string` and `str`, because `string` did strange concatenation. `listOf` and `attrsOf` have the same history with `attrs` and `list`, which are not able to recursively check their contents.

Compare the correspondence table below.